include/irr/asset/ITexturePacker.h

```diff
@@ -0,0 +1,483 @@
+#ifndef __IRR_C_TEXTURE_PACKER_H_INCLUDED__
+#define __IRR_C_TEXTURE_PACKER_H_INCLUDED__
+
+#include "irr/asset/format/EFormat.h"
+#include "irr/asset/ICPUImage.h"
+#include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+#include "irr/core/alloc/PoolAddressAllocator.h"
+#include "irr/core/math/morton.h"
+#include "irr/core/alloc/address_allocator_traits.h"
+
+namespace irr {
+namespace asset
+{
+
+class ITexturePacker
+{
+protected:
+    const uint32_t ADDR_LAYER_SHIFT;
+    const uint32_t ADDR_MORTON_MASK;
+    const uint32_t PHYS_PG_OFFSET_X_MASK;
+
+    const uint32_t m_pgSzxy;
+    const uint32_t m_tilesPerDim;
+
+    using pg_tab_addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
+    core::smart_refctd_dynamic_array<uint8_t> m_pgTabAddrAlctr_reservedSpc;
+    pg_tab_addr_alctr_t m_pgTabAddrAlctr;
+
+public:
+    //! SPhysPgOffset is what is stored in texels of page table!
+    struct SPhysPgOffset
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = 0xffffu;
+
+        SPhysPgOffset(uint32_t _addr) : addr(_addr) {}
+
+        //upper 16 bits are used for address of mip-tail page
+        uint32_t addr;
+    };
+    SPhysPgOffset makeTexOffset(uint32_t _x, uint32_t _y, uint32_t _layer)
+    {
+        uint32_t addr = (_x & PHYS_PG_OFFSET_X_MASK);
+        addr |= (_y & PHYS_PG_OFFSET_X_MASK) << (ADDR_LAYER_SHIFT >> 1);
+        addr |= (_layer << ADDR_LAYER_SHIFT);
+        addr |= (SPhysPgOffset::invalid_addr << 16);
+        return addr;
+    }
+    uint32_t physPgOffset_x(SPhysPgOffset _offset) const { return _offset.addr & PHYS_PG_OFFSET_X_MASK; }
+    uint32_t physPgOffset_y(SPhysPgOffset _offset) const { return (_offset.addr >> (ADDR_LAYER_SHIFT>>1)) & PHYS_PG_OFFSET_X_MASK; }
+    uint32_t physPgOffset_layer(SPhysPgOffset _offset) const { return (_offset.addr & 0xffffu)>>ADDR_LAYER_SHIFT; }
+    bool physPgOffset_valid(SPhysPgOffset _offset) const { return (_offset.addr&0xffffu) != SPhysPgOffset::invalid_addr; }
+    bool physPgOffset_hasMipTailAddr(SPhysPgOffset _offset) const { return physPgOffset_valid(physPgOffset_mipTailAddr(_offset)); }
+    SPhysPgOffset physPgOffset_mipTailAddr(SPhysPgOffset _offset) const { return _offset.addr>>16; }
+
+
+    using page_tab_offset_t = core::vector2du32_SIMD;
+    static page_tab_offset_t page_tab_offset_invalid() { return page_tab_offset_t(~0u,~0u); }
+
+    ITexturePacker(uint32_t _pgTabSzxy, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u) :
```

Comment:
I'd just take saner input parameters, such as log2 of the sizes (so then you don't need to assert for PoT)